hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java

```diff
@@ -16,35 +16,553 @@
+import com.hedera.node.app.hapi.utils.throttles.GasLimitDeterministicThrottle;
+import com.hedera.node.app.service.mono.throttling.ThrottleReqsManager;
+import com.hedera.node.app.service.token.ReadableAccountStore;
+import com.hedera.node.app.state.HederaState;
+import com.hedera.node.app.workflows.TransactionInfo;
+import com.hedera.node.app.workflows.dispatcher.ReadableStoreFactory;
+import com.hedera.node.config.ConfigProvider;
+import com.hedera.node.config.data.AccountsConfig;
+import com.hedera.node.config.data.AutoCreationConfig;
+import com.hedera.node.config.data.ContractsConfig;
+import com.hedera.node.config.data.LazyCreationConfig;
+import com.hedera.node.config.data.TokensConfig;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.config.api.Configuration;
 import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.math.BigInteger;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.EnumMap;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.IntSupplier;
+import javax.inject.Singleton;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 /**
- * Keeps track of the amount of usage of different throttle categories (by {@code id}), and returns
- * whether the throttle has been exceeded after applying the given incremental amount.
+ * Keeps track of the amount of usage of different TPS throttle categories and gas, and returns whether a given
+ * transaction or query should be throttled based on that.
+ * Meant to be used in single-threaded context only as part of the {@link com.hedera.node.app.workflows.handle.HandleWorkflow}.
  */
-public interface ThrottleAccumulator {
+@Singleton
+public class ThrottleAccumulator {
+
+    private static final Logger log = LogManager.getLogger(ThrottleAccumulator.class);
+    private static final Set<HederaFunctionality> GAS_THROTTLED_FUNCTIONS =
+            EnumSet.of(CONTRACT_CALL_LOCAL, CONTRACT_CALL, CONTRACT_CREATE, ETHEREUM_TRANSACTION);
+    private int capacitySplit;
+    private final ConfigProvider configProvider;
+    private EnumMap<HederaFunctionality, ThrottleReqsManager> functionReqs = new EnumMap<>(HederaFunctionality.class);
+    private static final int UNKNOWN_NUM_IMPLICIT_CREATIONS = -1;
+    private boolean lastTxnWasGasThrottled;
+    private GasLimitDeterministicThrottle gasThrottle;
+    private List<DeterministicThrottle> activeThrottles = Collections.emptyList();
+
+    public ThrottleAccumulator(
+            @NonNull final IntSupplier capacitySplitSource, @NonNull final ConfigProvider configProvider) {
+        this.configProvider = requireNonNull(configProvider, "configProvider must not be null");
+        this.capacitySplit = capacitySplitSource.getAsInt();
```

Comment:
Why do you use an `IntSupplier` instead of passing the `int` directly?

Reply:
Just to keep it consistent with the mono implementation(DeterministicThrottling). I can change it with **int** if you think it's better.

Reply:
I think the idea is when we initialize the frontend throttle like this
```
this.frontendThrottle =
                new ThrottleAccumulator(() -> platform.getAddressBook().getSize(), configProvider);
```
is to be able to get the address book size dynamically and use it for the capacity split but here we get the capacity split value only in the constructor which seems like something we should fix actually.

In mono we use the `IntSupplier` in [rebuildFor()](https://github.com/hashgraph/hedera-services/blob/06b8d8ae123ece1ab72c35f43c1097d1dfb0dc9b/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/throttling/DeterministicThrottling.java#L197) so it should pick up the address books size dynamically.

Reply:
That makes sense! I will fix that. Thanks

Reply:
@MiroslavGatsanoga fixed